### PR TITLE
[promxy] Header based query sharding

### PR DESCRIPTION
**Pull request by rishabhkumar92**
### Summary

This change is meant to support query sharding using headers. Currently headers which are allowlisted via proxy-headers are being passed to each server group and each target within a server group as it is. 

This leads to couple of problems on the backend:

- In Grafana Mimir's [context](https://grafana.com/docs/mimir/latest/manage/secure/authentication-and-authorization/#grafana-mimir-authentication-and-authorization), for federation usecases this can leads to lot of tenants in a single query and since Query layer on Mimir doesn't shard based on # of tenants, it leads to resource exhaustion and timeouts on storage system.
- Slower queries as single sharded query deals with lot of tenants under the hood.

This change adds support to split a query into multiple sub-queries, No of tenants send to backend is governed by batch-size parameter. If batch-size is not provided, behavior works as expected.

### Testing

- Tested query and label API with new behavior and ensure backward compatibility without new params.

cc: @aallawala @edma2

### CODE REVIEW

File: cmd/promxy/config.yaml
Comment (aallawala) [inline cmd/promxy/config.yaml:1]: changes in this file aren't meant to be committed, right? We can adjust those manually in our running version of promxy.

File: pkg/middleware/headers.go
Comment (aallawala) [inline pkg/middleware/headers.go:1]: can you add some tests here to validate the behavior?

File: pkg/promhttputil/merge.go [hunk 1]
```diff
@@ -171,6 +171,93 @@ func MergeValues(antiAffinityBuffer model.Time, a, b model.Value, preferMax bool
 	return nil, fmt.Errorf("unknown type! %v", reflect.TypeOf(a))
 }
 
+// SumValues sums values `a` and `b` with the given antiAffinityBuffer
+func SumValues(antiAffinityBuffer model.Time, a, b model.Value, preferMax bool) (model.Value, error) {
```
Comment (aallawala) [inline pkg/promhttputil/merge.go:175]: why is this needed as opposed to reusing promxy's constructs of combining subqueries when it issues it across multiple server groups?

Comment (rishabhkumar92) [reply pkg/promhttputil/merge.go:175]: This only differs for instant query where results are merged for same vector from different tenant.
When promxy combine subqueries from different server groups, it actually take max in vector.

**Comment by jacksontj** [comment]
So I have read the summary a few times and its a bit unclear to me the use case; so I'll attempt to summarize.

It sounds like when a query shows up to `promxy` it needs to aggregate across N different `tentants` within mimir. Today promxy sends a single query downstream -- causing issues; but the preference would be to have promxy "split" the query by tenant?

Assuming thats an accurate summarization of the feature request -- I *think* there is an easier way to achieve this; but I also may be completely missing the point (Haven't used mimir in a multi-tenant configuration with issues like this)

**Comment by rishabhkumar92** [comment]
> It sounds like when a query shows up to `promxy` it needs to aggregate across N different `tentants` within mimir. Today promxy sends a single query downstream -- causing issues; but the preference would be to have promxy "split" the query by tenant?

That's right, for a specific server group promxy send all information in header to Grafana Mimir which lacks splitting it and eventually queriers(worker) are doing lot of work per query. So we essentially want a way to split them upfront and batch them before sending it to server groups.

> I think there is an easier way to achieve this

please feel free to leave a suggestion.

**Comment by jacksontj** [comment]
The first approach that jumps to mind is having promxy configured with each tenant as its own servergroup. Then each request to that servergroup would have the appropriate headers (and if you needed to query across all N it would split queries and aggregate results).

If there are a lot of tenants this may become burdensome for configuration -- but if you can get it in a k8s label or some other relabel_config supported mechanism -- there are some tricks to "template" that part out as well (I'll have to see if I can find an example of that somewhere).

Hopefully that is helpful and/or makes sense? Basically if you want them treated as distinct endpoints -- then we would ideally have them configured as distinct servergroups (which seems like it would cover your use-case).

**Comment by rishabhkumar92** [comment]
> If there are a lot of tenants this may become burdensome for configuration -- but if you can get it in a k8s label or some other relabel_config supported mechanism -- there are some tricks to "template" that part out as well (I'll have to see if I can find an example of that somewhere).

So performance is the main reason why I want to split these requests into batches of tenants. one servergroup per tenant sounds like a lot granular and might have significant impact on resource usage on promxy if it needs to deal with each tenant 's result separately and that was the reason to not go via that route. Dealing with 1000s of tenant here and batch size of 50/100 works out to be a good balance in term of promxy resource usage and query performance.

**Comment by jacksontj** [comment]
> one servergroup per tenant sounds like a lot granular and might have significant impact on resource usage on promxy if it needs to deal with each tenant 's result separately

this would depend on the query load. in *some* ways it would be easier as the data can be deduped in smaller batches -- but if you are constantly querying all or most -- then it would require N queries to be merged instead of a smaller number downstream.

> Dealing with 1000s of tenant here and batch size of 50/100 works out to be a good balance in term of promxy resource usage and query performance.

So if I'm understanding this correctly; basically there is *a* header that comes from upstream (grafana) with a list of 1k tennants -- and you want to effectively split/shard the query based on some number of those. Being a bit more specific we don't need them to be split 1by1 or consistently; just that any batch doesn't have more than N of those headers in there.


If so; then your approach is quite close to what I'd suggest here. I think it might be cleaner to implement another API "layer" that supports doing the splitting and combining (likely leveraging the multi-API under the hood). That *seems* like it would create less of a large change -- and would make the config a bit cleaner (just another API client, maybe leveraging the multiAPI?).

**Comment by pmatuszak** [comment]
+1 to implement this future :)

**Comment by rishabhkumar92** [comment]
I am closing this as I am currently working on upgrading Promxy on recent prometheus and add native histogram support.
